File: app/src/main/java/net/schueller/peertube/service/LoginService.java
/*
 * Copyright 2018 Stefan Schüller <sschueller@techdroid.com>
 *
 * License: GPL-3.0+
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package net.schueller.peertube.service;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;

import net.schueller.peertube.R;
import net.schueller.peertube.application.AppApplication;
import net.schueller.peertube.helper.APIUrlHelper;
import net.schueller.peertube.model.OauthClient;
import net.schueller.peertube.model.Token;
import net.schueller.peertube.network.AuthenticationService;
import net.schueller.peertube.network.RetrofitInstance;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static net.schueller.peertube.application.AppApplication.getContext;

public class LoginService {

    private static final String TAG = "LoginService";

    public static void Authenticate(String username, String password) {
        Context context = getContext();

        String apiBaseURL = APIUrlHelper.getUrlWithVersion(context);

        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(context);

        AuthenticationService service = RetrofitInstance.getRetrofitInstance(apiBaseURL, APIUrlHelper.useInsecureConnection(context)).create(AuthenticationService.class);

        Call<OauthClient> call = service.getOauthClientLocal();

        call.enqueue(new Callback<OauthClient>() {
            @Override
            public void onResponse(@NonNull Call<OauthClient> call, @NonNull Response<OauthClient> response) {

                if (response.isSuccessful()) {

                    OauthClient oauthClient = response.body();

                    SharedPreferences.Editor editor = sharedPref.edit();

                    assert oauthClient != null;

                    editor.putString(context.getString(R.string.pref_client_id), oauthClient.getClientId());
                    editor.putString(context.getString(R.string.pref_client_secret), oauthClient.getClientSecret());
                    editor.apply();

                    Call<Token> call2 = service.getAuthenticationToken(
                            oauthClient.getClientId(),
                            oauthClient.getClientSecret(),
                            "code",
                            "password",
                            "upload",
                            username,
                            password
                    );
                    call2.enqueue(new Callback<Token>() {
                        @Override
                        public void onResponse(@NonNull Call<Token> call2, @NonNull retrofit2.Response<Token> response2) {

                            if (response2.isSuccessful()) {

                                Token token = response2.body();

                                assert token != null;
                                editor.putString(context.getString(R.string.pref_token_access), token.getAccessToken());
                                editor.putString(context.getString(R.string.pref_token_refresh), token.getRefreshToken());
                                editor.putString(context.getString(R.string.pref_token_type), token.getTokenType());
                                editor.apply();

                                Log.wtf(TAG, "Logged in");

                                Toast.makeText(context, context.getString(R.string.authentication_login_success), Toast.LENGTH_LONG).show();


                            } else {
                                Log.wtf(TAG, response2.toString());
                                Toast.makeText(context, context.getString(R.string.authentication_login_failed), Toast.LENGTH_LONG).show();

                            }
                        }

                        @Override
                        public void onFailure(@NonNull Call<Token> call2, @NonNull Throwable t2) {
                            Log.wtf("err", t2.fillInStackTrace());
                            Toast.makeText(context, context.getString(R.string.authentication_login_failed), Toast.LENGTH_LONG).show();

                        }
                    });

                } else {
                    Log.wtf(TAG, response.toString());
                    Toast.makeText(context, context.getString(R.string.authentication_login_failed), Toast.LENGTH_LONG).show();

                }
            }

            @Override
            public void onFailure(@NonNull Call<OauthClient> call, @NonNull Throwable t) {
                Log.wtf("err", t.fillInStackTrace());
                Toast.makeText(context, context.getString(R.string.authentication_login_failed), Toast.LENGTH_LONG).show();

            }
        });
    }

    public static void refreshToken() {
        Context context = getContext();

        String apiBaseURL = APIUrlHelper.getUrlWithVersion(context);
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
        AuthenticationService service = RetrofitInstance.getRetrofitInstance(apiBaseURL, APIUrlHelper.useInsecureConnection(context)).create(AuthenticationService.class);

        String refreshToken = sharedPreferences.getString(AppApplication.getContext().getString(R.string.pref_token_refresh), null);
        String userName = sharedPreferences.getString(AppApplication.getContext().getString(R.string.pref_auth_username), null);
        String clientId = sharedPreferences.getString(AppApplication.getContext().getString(R.string.pref_client_id), null);
        String clientSecret = sharedPreferences.getString(AppApplication.getContext().getString(R.string.pref_client_secret), null);


        Call<Token> call = service.refreshToken(
                clientId,
                clientSecret,
                "refresh_token",
                "code",
                userName,
                refreshToken
        );
        call.enqueue(new Callback<Token>() {
            @Override
            public void onResponse(@NonNull Call<Token> call, @NonNull retrofit2.Response<Token> response) {

                if (response.isSuccessful()) {

                    Token token = response.body();

                    SharedPreferences.Editor editor = sharedPreferences.edit();

                    assert token != null;
                    editor.putString(context.getString(R.string.pref_token_access), token.getAccessToken());
                    editor.putString(context.getString(R.string.pref_token_refresh), token.getRefreshToken());
                    editor.putString(context.getString(R.string.pref_token_type), token.getTokenType());
                    editor.apply();

                    Log.wtf(TAG, "Logged in");

                    Toast.makeText(context, context.getString(R.string.authentication_token_refresh_success), Toast.LENGTH_LONG).show();


                } else {
                    Log.wtf(TAG, response.toString());
                    Toast.makeText(context, context.getString(R.string.authentication_token_refresh_failed), Toast.LENGTH_LONG).show();

                }
            }

            @Override
            public void onFailure(@NonNull Call<Token> call2, @NonNull Throwable t2) {
                Log.wtf("err", t2.fillInStackTrace());
                Toast.makeText(context, context.getString(R.string.authentication_token_refresh_failed), Toast.LENGTH_LONG).show();

            }
        });
    }
}

File: app/src/main/java/net/schueller/peertube/feature_video/data/remote/auth/LoginService.kt
package net.schueller.peertube.feature_video.data.remote.auth

import android.content.Context

import android.util.Log
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import net.schueller.peertube.common.Constants.PREF_AUTH_USERNAME
import net.schueller.peertube.common.Constants.PREF_CLIENT_ID
import net.schueller.peertube.common.Constants.PREF_CLIENT_SECRET
import net.schueller.peertube.common.Constants.PREF_TOKEN_ACCESS
import net.schueller.peertube.common.Constants.PREF_TOKEN_EXPIRATION
import net.schueller.peertube.common.Constants.PREF_TOKEN_REFRESH
import net.schueller.peertube.common.Constants.PREF_TOKEN_TYPE
import net.schueller.peertube.feature_video.data.remote.PeerTubeApi
import net.schueller.peertube.feature_video.data.remote.dto.OauthClientDto
import retrofit2.HttpException
import java.io.IOException
import javax.inject.Inject


class LoginService @Inject constructor(
    @ApplicationContext private val context: Context,
    private val api: PeerTubeApi
){
    private val tag = "LoginService"
    private val sharedPreferences = context.getSharedPreferences(context.packageName + "_preferences", Context.MODE_PRIVATE)

    fun authenticate(username: String?, password: String?) {

        try {
            CoroutineScope(Dispatchers.IO).launch {

                val oauthClientDto: OauthClientDto = api.getOauthClientLocal()

                // store new client id and secret
                val editor = sharedPreferences.edit()
                editor.putString(PREF_CLIENT_ID, oauthClientDto.clientId)
                editor.putString(PREF_CLIENT_SECRET, oauthClientDto.clientSecret)
                editor.apply()

                Log.wtf(tag, "sharedPreferences save")


                Log.wtf(tag, "getAuthenticationToken 1")


                Log.wtf(tag, "getAuthenticationToken 2")

                val response = api.getAuthenticationToken(
                    oauthClientDto.clientId,
                    oauthClientDto.clientSecret,
                    "code",
                    "password",
                    "upload",
                    username,
                    password
                )

                val tokenDto = response.body()

                if (response.isSuccessful && tokenDto != null) {

                    Log.wtf(tag, "accessToken: "  + tokenDto.accessToken)

                    editor.putString(PREF_TOKEN_ACCESS, tokenDto.accessToken)
                    editor.putString(PREF_TOKEN_REFRESH, tokenDto.refreshToken)
                    editor.putString(PREF_TOKEN_EXPIRATION, tokenDto.tokenType)

                    editor.apply()

                } else {
                    Log.wtf(tag, "Login failed")
                }

            }

        } catch (e: HttpException) {
            Log.wtf(tag, e.localizedMessage)
        } catch (e: IOException) {
            Log.wtf(tag, "Login Error")
        }

    }

    fun refreshToken() {

        try {

            val refreshToken = sharedPreferences.getString(PREF_TOKEN_REFRESH, null)
            val userName = sharedPreferences.getString(PREF_AUTH_USERNAME, null)
            val clientId = sharedPreferences.getString(PREF_CLIENT_ID, null)
            val clientSecret = sharedPreferences.getString(PREF_CLIENT_SECRET, null)

            CoroutineScope(Dispatchers.IO).launch {

                val response = api.refreshToken(
                    clientId,
                    clientSecret,
                    "refresh_token",
                    "code",
                    userName,
                    refreshToken
                )

                val tokenDto = response.body()

                if (response.isSuccessful && tokenDto != null) {
                    val editor = sharedPreferences.edit()
                    editor.putString(PREF_TOKEN_ACCESS, tokenDto.accessToken)
                    editor.putString(PREF_TOKEN_REFRESH, tokenDto.refreshToken)
                    editor.putString(PREF_TOKEN_TYPE, tokenDto.tokenType)
                    editor.apply()
                    Log.wtf(tag, "Logged in")
                } else {
                    Log.wtf(tag, "Login failed")
                }

            }

        }  catch (e: HttpException) {
            Log.wtf(tag, e.localizedMessage)
        } catch (e: IOException) {
            Log.wtf(tag, "Refresh Error")
        }

    }

}